haskell/test/SMT/Model/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,368 @@
+      else do
+        slashAmount <- rejectedProposalSlash -$$? (proposal, (store & sExtra))
+        let frozenTokens = (proposal & plProposerFrozenToken) + fixedFee
+            desiredBurnAmount = arg #slash_amount slashAmount + fixedFee
+            tokens = case frozenTokens `minusNaturalMaybe` desiredBurnAmount of
+              Just value -> value
+              Nothing -> 0
+        pure (tokens, desiredBurnAmount)
+
+  unstakeTk tokens burnAmount (proposal & plProposer)
+
+  mapM_ doUnfreeze (Map.toList (proposal & plVoters))
+  where
+    doUnfreeze :: ((Address, Bool), Natural) -> ModelT ()
+    doUnfreeze ((addr, _), voteAmount) = do
+      unstakeTk voteAmount 0 addr
+
+deleteProposal :: Natural -> ProposalKey -> ModelT ()
+deleteProposal pLevel proposalKey = do
+  modifyStore $ \s -> s { sProposalKeyListSortByDate = Set.delete (pLevel, proposalKey) (s & sProposalKeyListSortByDate)}
+
+applyPropose :: ModelCall -> ProposeParams -> ModelT ()
+applyPropose mc param@ProposeParams{..} = do
+
+  config <- getConfig
+  let FixedFee fixedFee = config & cFixedProposalFee
+
+  validFrom <- checkDelegate ppFrom (mc & mcSender)
+
+  store <- getStore
+  _    <- (config & cProposalCheck) -$$? (param, (store & sExtra))
+  checkProposalLimitReached
+
+  let amountToFreeze = ppFrozenToken + fixedFee
+  currentStage <- getCurrentStageNum
+
+  updateQuorum currentStage
+  stakeTk amountToFreeze validFrom
+  addProposal param
+
+checkVoteLimitReached :: Proposal -> ModelT ()
+checkVoteLimitReached pl = do
+  maxVoters <- getConfig <&> cMaxVoters
+  when (maxVoters <= (fromIntegral $ length (pl & plVoters))) $
+    throwError MAX_VOTERS_REACHED
+
+ensureProposalVotingStage :: Proposal -> ModelT ()
+ensureProposalVotingStage proposal = do
+  currentStage <- getCurrentStageNum
+  unless (currentStage == (proposal & plVotingStageNum)) $
+    throwError VOTING_STAGE_OVER
+
+submitVote :: Proposal -> VoteParam -> Address -> ModelT ()
+submitVote proposal voteParam author = do
+  let proposalKey = voteParam & vProposalKey
+      proposal_ =
+        let mapKey = (author, voteParam & vVoteType)
+            newVotes = case Map.lookup mapKey (proposal & plVoters) of
+              Just votes -> votes + (voteParam & vVoteAmount)
+              Nothing -> voteParam & vVoteAmount
+        in proposal { plVoters = Map.insert mapKey newVotes (proposal & plVoters) }
+
+      updatedProposal = if (voteParam & vVoteType)
+                        then proposal_ { plUpvotes = (proposal_ & plUpvotes) + (voteParam & vVoteAmount)}
+                        else proposal_ { plDownvotes = (proposal_ & plDownvotes) + (voteParam & vVoteAmount)}
+  stakeTk (voteParam & vVoteAmount) author
+  modifyStore $ \s -> s { sProposals = BigMap $ Map.insert proposalKey updatedProposal (s & sProposals & T.unBigMap)}
+
+
+applyVote :: ModelCall -> [PermitProtected VoteParam] -> ModelT ()
+applyVote ms params = do
```

Comment:
I suggest here to only do
```suggestion
applyVote :: ModelCall -> [PermitProtected VoteParam] -> ModelT ()
applyVote ms = mapM_ submitVote
```
and move the checks of `acceptVote` into `submitVote` as well.

Reply:
Fixed.